core/src/main/java/io/grpc/internal/ServiceConfigUtil.java

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2018, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Verify;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Random;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * Helper utility to work with service configs.
+ */
+final class ServiceConfigUtil {
+
+  private static final Logger logger = Logger.getLogger(ServiceConfigUtil.class.getName());
+
+  // From https://github.com/grpc/proposal/blob/master/A2-service-configs-in-dns.md
+  static final String SERVICE_CONFIG_PREFIX = "_grpc_config=";
+  private static final Set<String> SERVICE_CONFIG_CHOICE_KEYS =
+      Collections.unmodifiableSet(
+          new HashSet<String>(
+              Arrays.asList("clientLanguage", "percentage", "clientHostname", "serviceConfig")));
```

Comment:
I would prefer to leave them as is.   It is much easier to see whats in the list without having to go through an indirection.

Reply:
They are reused in multiple places and it's hard to find the bug if one of them happens to be a typo. It is also hard to refactor in future because there are multiple strings with the same value but for different usage, like "serviceConfig", so it's hard to tell which should be refactored together and which should not.

Reply:
Since the values are part of the API, they cannot change.   This is going to get out of hand when all the other fields are added in, and they cannot match the proto which is how the data will be set.   

I would strong prefer to keep these as is and not add another level of indirection for minimal payoff.

Reply:
- Declaring the constants just one line above the set is not that indirect.
- You already had another indirect for `SERVICE_CONFIG_PREFIX` and should follow suite.
- There are multiple occurrence of "percentage" etc and you have to make sure each of them is spelled correctly otherwise a typo might hide a bug. In contrast, if you declare it as a constant, then you just make sure in one place the spelling is correct, and other places just don't care the spelling but only the meaning of the variable. The variable name PERCENTAGE is clear in meaning.

Reply:
Acknowledged.  If this gets out of hand, I will move them over to constants.  There are many more coming values going to be put here.